File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for(int i=0;i<s.length();i++)
        {
            char ch = s[i];

            if(ch == '(' || ch == '{' || ch == '[')
            {
                st.push(ch);
            }
            //now check close parenthesis
           else
           {
            //always remember to check empty stack or not
                if(ch == ')' && !st.empty() && st.top() == '(')
                {
                    st.pop();
                }

                else if(ch == '}' && !st.empty() && st.top() == '{')
                {
                    st.pop();
                }

                else if(ch == ']' && !st.empty() && st.top() == '[')
                {
                    st.pop();
                }

                else{
                    return false;
                }
           }

        }
        // check if it is empty or not
        if(st.size() == 0)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {

            // 1. If opening bracket → push
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                // 2. If closing bracket → stack must not be empty
                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();  // latest opening bracket

                // 3. Check if types match
                if (c == ')' && top != '(') return false;
                if (c == '}' && top != '{') return false;
                if (c == ']' && top != '[') return false;
            }
        }

        // 4. All opening brackets must be matched
        return stack.isEmpty();
    }
}